### PR TITLE
AVRO-3601: C++ API header contains breaking include

### PR DESCRIPTION
**Pull request by martin-g**
Rename CustomFields to CustomAttributes.
Rework CustomAttributes to keep a map of string name with a string
value. The user application may parse the string to JSON if needed.
Add a new step to the CI workflow to prevent problems like this in the
future

### Jira

- [X] https://issues.apache.org/jira/browse/AVRO-3601

### Tests

- [X] My PR updates the relevant unit tests

### Commits

- [X] My commits all reference Jira issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](https://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain Javadoc that explain what it does

### CODE REVIEW

**Comment by martin-g** [comment]
CC @nileyadav @mruffing @chenrui333

**Review by thiru-mg — APPROVED** [review]
Looks great! Thank you for the prompt action.

File: lang/c++/impl/CustomAttributes.cc [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+#include <memory>
+#include "Exception.hh"
+
+namespace avro {
+
+std::string CustomAttributes::getAttribute(const std::string &name) const {
+    std::map<std::string, std::string>::const_iterator iter =
+        attributes_.find(name);
+    if (iter == attributes_.end()) {
+      return NULL;
```
[the remark this answers is not in the document]
Comment (martin-g) [reply lang/c++/impl/CustomAttributes.cc:30]: No, there is no such test.
Suggestions how to rework this ? Maybe return Option(al) ?! I have to see which version of C++ supports this.

File: lang/c++/impl/Compiler.cc [hunk 2]
```diff
@@ -268,14 +268,14 @@ static const std::unordered_set<std::string>& getKnownFields() {
...
-static void getCustomAttributes(const Object& m, CustomFields &customAttributes)
+static void getCustomAttributes(const Object& m, CustomAttributes &customAttributes)
 {
   // Don't add known fields on primitive type and fixed type into custom
   // fields.
   const std::unordered_set<std::string>& kKnownFields = getKnownFields();
   for (const auto &entry : m) {
     if (kKnownFields.find(entry.first) == kKnownFields.end()) {
-      customAttributes.addField(entry.first, entry.second);
+      customAttributes.addAttribute(entry.first, entry.second.stringValue());
```
[the remark this answers is not in the document]
Comment (martin-g) [reply lang/c++/impl/Compiler.cc:278]: I think the C++ developers here have to step up! :-)
Thanks for the review, @KalleOlaviNiemitalo !

File: lang/c++/impl/CustomAttributes.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (martin-g) [reply lang/c++/impl/CustomAttributes.cc:30]: `std::optional` is available since C++ 17. Avro uses 11

[the remark this answers is not in the document]
Comment (martin-g) [reply lang/c++/impl/CustomAttributes.cc:30]: `boost::optional` is available!

[the remark this answers is not in the document]
Comment (martin-g) [reply lang/c++/impl/CustomAttributes.cc:30]: https://github.com/apache/avro/pull/1826

File: lang/c++/impl/Compiler.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (martin-g) [reply lang/c++/impl/Compiler.cc:278]: https://github.com/apache/avro/pull/1826/commits/074eef4272bf563a334e42a1027890f0c34dcccc
Looks good ?

**Comment by nileyadav** [comment]
Thank you @martin-g for promptly fixing this.
Will the CustomAttribute feature and this fix be part of release?

Regards,
Nilesh

**Comment by martin-g** [comment]
It will be part of 1.11.2 and 1.12.0 when they are released.
There are no discussions about specific dates for now.

**Pull request merged**